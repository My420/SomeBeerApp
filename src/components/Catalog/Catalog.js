import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';
import { connect } from 'react-redux';
import ProductList from '../ProductList/ProductList';
import GridItemCard from '../GridItemCard/GridItemCard';
import IconLoadingSvg from '../IconLoadingSvg/IconLoadingSvg';
import ErrorMsg from '../ErrorMsg/ErrorMsg';
import styles from './Catalog.module.scss';
import getCatalogBeerData from '../../ActionCreator/getCatalogBeerData';
import CatalogFilter from '../CatalogFilter/CatalogFilter';
import LoadStatusSwitcher from '../LoadStatusSwitcher/LoadStatusSwitcher';
import CatalogPagination from '../CatalogPagination/CatalogPagination';
import {
  ABV_LESS_PROP,
  ABV_MORE_PROP,
  IBU_LESS_PROP,
  IBU_MORE_PROP,
  EBC_LESS_PROP,
  EBC_MORE_PROP,
  PAGE,
  BEER_NAME_PROP
} from '../../utils/constants';

export class Catalog extends React.Component {
  componentDidMount() {
    const { getCatalogData, options } = this.props;
    getCatalogData(options);
  }

  componentDidUpdate(prevProps) {
    const { options: prevOpt } = prevProps;
    const { options: currentOpt } = this.props;
    if (prevOpt !== currentOpt) {
      const { getCatalogData, options } = this.props;
      getCatalogData(options);
    }
  }

  render() {
    // eslint-disable-next-line no-console
    console.log('render ========== Catalog');
    const {
      history,
      options,
      isLoaded,
      isLoading,
      isError,
      errorMessage,
      data
    } = this.props;

    const isDataEmpty = data.size === 0;

    const LoadingComponent = () => {
      return (
        <div className={styles.loader}>
          <IconLoadingSvg color="#e69e63" />
        </div>
      );
    };

    const ErrorComponent = () => {
      return (
        <div className={styles.error}>
          <ErrorMsg errorMsg={errorMessage} />
        </div>
      );
    };

    const DataComponent = () => {
      return <ProductList data={data} ProductCard={GridItemCard} />;
    };

    return (
      <React.Fragment>
        <CatalogFilter value={options} history={history} />
        <LoadStatusSwitcher
          isLoading={isLoading}
          isError={isError}
          isLoaded={isLoaded}
          LoadingComponent={LoadingComponent}
          ErrorComponent={ErrorComponent}
          DataComponent={DataComponent}
        />
        {!isDataEmpty && <CatalogPagination options={options} />}
      </React.Fragment>
    );
  }
}

const mapStateToProps = store => {
  return {
    isLoaded: store.catalog.isLoaded,
    isLoading: store.catalog.isLoading,
    isError: store.catalog.isError,
    errorMessage: store.catalog.errorMessage,
    data: store.catalog.data
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getCatalogData: options => {
      dispatch(getCatalogBeerData(options));
    }
  };
};

Catalog.propTypes = {
  history: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  options: PropTypes.shape({
    [PAGE]: PropTypes.string.isRequired,
    [BEER_NAME_PROP]: PropTypes.string.isRequired,
    [ABV_LESS_PROP]: PropTypes.string.isRequired,
    [ABV_MORE_PROP]: PropTypes.string.isRequired,
    [IBU_LESS_PROP]: PropTypes.string.isRequired,
    [IBU_MORE_PROP]: PropTypes.string.isRequired,
    [EBC_LESS_PROP]: PropTypes.string.isRequired,
    [EBC_MORE_PROP]: PropTypes.string.isRequired
  }).isRequired,
  isLoaded: PropTypes.bool.isRequired, // form connect
  isLoading: PropTypes.bool.isRequired, // form connect
  isError: PropTypes.bool.isRequired, // form connect
  errorMessage: PropTypes.string.isRequired, // form connect
  data: PropTypes.instanceOf(List).isRequired, // form connect
  getCatalogData: PropTypes.func.isRequired // form connect
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Catalog);
