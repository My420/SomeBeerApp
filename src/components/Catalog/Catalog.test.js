import React from 'react';
import { shallow } from 'enzyme';
import { Catalog } from './Catalog';

describe('test <Catalog /> component', () => {
  const options = { name: 'test' };
  const history = { path: 'test' };
  const other = {
    isLoading: true,
    isLoaded: false,
    isError: false,
    errorMessage: '',
    data: { test: 'test' }
  };

  const mockFunc = jest.fn();

  const wrapper = shallow(
    <Catalog
      options={options}
      history={history}
      {...other}
      getCatalogData={mockFunc}
    />
  );

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should render CatalogFilter component with correct props', () => {
    expect(wrapper.find('CatalogFilter')).toHaveLength(1);
    expect(wrapper.find('CatalogFilter').prop('history')).toEqual(history);
    expect(wrapper.find('CatalogFilter').prop('value')).toEqual(options);
  });

  test('should render LoadStatusSwitcher component with correct props', () => {
    const { isLoaded, isLoading, isError } = other;
    expect(wrapper.find('LoadStatusSwitcher')).toHaveLength(1);
    expect(wrapper.find('LoadStatusSwitcher').prop('isLoading')).toBe(
      isLoading
    );
    expect(wrapper.find('LoadStatusSwitcher').prop('isLoaded')).toBe(isLoaded);
    expect(wrapper.find('LoadStatusSwitcher').prop('isError')).toBe(isError);
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('LoadingComponent')
    ).toBe('function');
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('ErrorComponent')
    ).toBe('function');
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('DataComponent')
    ).toBe('function');
  });

  test('should call getCatalogData function in componentDidMount', () => {
    expect(mockFunc.mock.calls.length).toBe(1);
    expect(mockFunc.mock.calls[0][0]).toEqual(options);
  });

  test('should not call getCatalogData in compDidUpdate if prop options is same', () => {
    wrapper.setProps({ options });
    expect(mockFunc.mock.calls.length).toBe(1);
  });

  test('should  call getCatalogData in compDidUpdate if prop option is not same', () => {
    wrapper.setProps({ options: { name: 'test2' } });
    expect(mockFunc.mock.calls.length).toBe(2);
    expect(mockFunc.mock.calls[1][0]).toEqual({ name: 'test2' });
  });

  test('should  not render CatalogPagination component if prop data.size = 0', () => {
    wrapper.setProps({ data: { size: 0 } });
    expect(wrapper.find('CatalogPagination')).toHaveLength(0);
  });

  test('should  render CatalogPagination component if prop data.size != 0', () => {
    wrapper.setProps({ data: { size: 33 }, options: { name: 'test2' } });
    expect(wrapper.find('CatalogPagination')).toHaveLength(1);
  });
});
