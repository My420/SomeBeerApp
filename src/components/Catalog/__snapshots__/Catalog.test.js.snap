// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <Catalog /> component should match snapshot 1`] = `
<Fragment>
  <CatalogFilter
    history={
      Object {
        "path": "test",
      }
    }
    value={
      Object {
        "name": "test",
      }
    }
  />
  <LoadStatusSwitcher
    DataComponent={[Function]}
    ErrorComponent={[Function]}
    LoadingComponent={[Function]}
    isError={false}
    isLoaded={false}
    isLoading={true}
  />
  <CatalogPagination
    options={
      Object {
        "name": "test",
      }
    }
  />
</Fragment>
`;
